kingpin/actors/aws/s3.py

```diff
@@ -178,48 +178,117 @@ class LifecycleConfig(SchemaCompareBase):
+                        {
+                            'required': 'filter'
+                        },
+                        {
+                            'required': 'prefix'
+                        }
+                    ]
+                },
+                {
+                    'anyOf': [
+                        {
+                            'required': 'transition'
+                        },
+                        {
+                            'required': 'noncurrent_version_transition'
+                        },
+                        {
+                            'required': 'expiration'
+                        },
+                        {
+                            'required': 'noncurrent_version_expiration'
+                        },
+                        {
+                            'required': 'abort_incomplete_multipart_upload'
+                        }
+                    ]
+                }
+            ],
             'additionalProperties': False,
             'properties': {
-                # The ID and Prefix must be strings. We do not allow for them
-                # to be empty -- they must be defined.
+                # The ID must be a string. We do not allow for it to be
+                # empty -- it must be defined.
                 'id': {
                     'type': 'string',
                     'minLength': 1,
-                    'maxLength': 255,
-                },
-                'prefix': {
-                    'type': 'string',
+                    'maxLength': 255
                 },
 
-                # The Status field must be 'Enabled' or 'Disabled'
+               # The Status field must be 'Enabled' or 'Disabled'
                 'status': {
                     'type': 'string',
-                    'enum': ['Enabled', 'Disabled'],
+                    'enum': ['Enabled', 'Disabled']
                 },
 
-                # Expiration and Transition can be empty, or have
-                # configurations associated with them.
-                'expiration': {
-                    'type': ['string', 'integer', 'object'],
-                    'pattern': '^[0-9]+$',
+                # Filter is a new way of defining rule prefixes
+                'filter': {
+                    'type': 'object',
+                    'minProperties': 1,
+                    'maxProperties': 1,
                     'additionalProperties': False,
                     'properties': {
-                        'days': {
-                            'type': ['string', 'integer'],
-                            'pattern': '^[0-9]+$',
-                        },
-                        'date': {
+                        'prefix': {
                             'type': 'string',
-                            'format': 'date-time',
                         },
-                        'expired_object_delete_marker': {
-                            'type': 'boolean',
+                        'tag': {
```

Comment:
You can probably move this out to a tag object, and reference it there, since you're playing with python dicts. That way you can also reuse this down in the other blocks

Reply:
Good call - there's actually a proper syntax in JsonSchema where you create a Tag object and just refer to it by name. You move this into the `definitions` key at the top of the schema. Then you can use `$ref` to refer to it... kinda like this:
```
"tag": { "$ref": "#/definitions/Tag" }
```

Reply:
yup saw that, learning json schema on the fly :) I will switch to using a definition